File: Problems/Tic Tac Toe/java/tictactoe/TicTacToe.java
package tictactoe;

import board.Board;
import player.Player;
import java.util.Stack;

public class TicTacToe {
    private static volatile TicTacToe instance;
    private Board board;
    private Player[] players;
    private Stack<Player> moves;
    private GameState gameState;

    private TicTacToe(int size, Player[] players){
        this.board = new Board(size);
        this.players = new Player[2];
        this.players = players;
        gameState = GameState.IN_PROGRESS;
        moves = new Stack<>();
    }

    public static synchronized TicTacToe getTicTacToeInstance(int size, Player[] players) {
        if(instance == null){
            instance = new TicTacToe(size, players);
        }

        return instance;
    }

    public Player[] getPlayers() {
        return players;
    }

    public void playMove(int x, int y, Player player) throws Exception{
        if(gameState != GameState.IN_PROGRESS){
            throw new Exception("Game is already over");
        }

        if(board.isValidMove(x, y, player.getShape())){
            if(moves.isEmpty() || moves.peek() != player){
                board.playMove(x, y, player.getShape());
                moves.push(player);
                board.printBoard();
                // check for winning conditions
                if(board.isWinner(player.getShape())){
                    gameState = GameState.WIN;
                    System.out.println("Player " + player.getName() + " is the winner !!!");
                    return;
                }

                // check for draw condition
                if(board.isDraw(moves.size())){
                    gameState = GameState.DRAW;
                    System.out.println("DRAW");
                    return;
                }
            } else {
                throw new Exception("Wrong Player move");
            }
        } else {
            throw new Exception("Invalid Move");
        }
    }
    
}

File: Problems/Tic Tac Toe/Golang/tictac/tictactoe.go
package tictac

import (
	"errors"
	"fmt"
	"sync"
	"tictactoe/board"
	"tictactoe/player"
)

// TicTacToe represents the game logic.
type TicTacToe struct {
	board     *board.Board
	players   []*player.Player
	moves     []*player.Player
	gameState GameState
	// Mutex for concurrent access
	mu sync.Mutex
}

// Instance for the singleton pattern.
var (
	instance *TicTacToe
	once     sync.Once
)

// GetTicTacToeInstance implements the singleton pattern for TicTacToe.
func GetTicTacToeInstance(size int, players []*player.Player) *TicTacToe {
	once.Do(func() {
		instance = &TicTacToe{
			board:     board.NewBoard(size),
			players:   players,
			moves:     make([]*player.Player, 0), // Initialize an empty slice
			gameState: IN_PROGRESS,
		}
	})
	return instance
}

// GetPlayers returns the players of the game.
func (t *TicTacToe) GetPlayers() []*player.Player {
	return t.players
}

// PlayMove handles a player's move.
func (t *TicTacToe) PlayMove(x, y int, player *player.Player) error {
	// Lock to prevent concurrent moves
	t.mu.Lock()
	defer t.mu.Unlock()

	if t.gameState != IN_PROGRESS {
		return errors.New("game is already over")
	}

	// Check if it's the correct player's turn
	if len(t.moves) > 0 && t.moves[len(t.moves)-1] == player {
		return errors.New("wrong player move")
	}

	if t.board.IsValidMove(x, y) {
		t.board.PlayMove(x, y, player.Shape)
		t.moves = append(t.moves, player)
		t.board.PrintBoard()

		// Check for winning conditions
		if t.board.IsWinner(player.Shape) {
			t.gameState = WIN
			fmt.Printf("Player %s is the winner !!!\n", player.Name)
			return nil
		}

		// Check for draw condition
		if t.board.IsDraw(len(t.moves)) {
			t.gameState = DRAW
			fmt.Println("DRAW")
			return nil
		}

		return nil
	} else {
		return errors.New("invalid move")
	}
}
